sdk/helper/nonce/encrypted_nonce.go

```diff
@@ -0,0 +1,374 @@
+	ensCounter   uint64
+)
+
+type encryptedNonceService struct {
+	validity time.Duration
+
+	crypt cipher.AEAD
+
+	nextCounter *atomic.Uint64
+
+	issueLock      *sync.Mutex
+	maxIssued      map[ensTimestamp]ensCounter
+	minCounter     *atomic.Uint64
+	redeemedTokens map[ensTimestamp]map[ensCounter]struct{}
+}
+
+func newEncryptedNonceService(validity time.Duration) *encryptedNonceService {
+	return &encryptedNonceService{
+		validity: validity,
+		// nextCounter.Add(1) returns the _new_ value; by initializing to
+		// zero, we guarantee that nextCounter = minCounter + 1 on the first
+		// read; if it is redeemed right away, we then hold that invariant.
+		nextCounter: new(atomic.Uint64),
+
+		issueLock:      new(sync.Mutex),
+		maxIssued:      make(map[ensTimestamp]ensCounter, validity/time.Second),
+		minCounter:     new(atomic.Uint64),
+		redeemedTokens: make(map[ensTimestamp]map[ensCounter]struct{}, validity/time.Second),
+	}
+}
+
+func (ens *encryptedNonceService) Initialize() error {
+	// On initialization, create a new AES key. This avoids having issues
+	// with the number of encryptions we can do under this service.
+	key := make([]byte, 32)
+	if _, err := io.ReadFull(rand.Reader, key); err != nil {
+		return fmt.Errorf("failed to initialize AES key: %w", err)
+	}
+
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return fmt.Errorf("failed to initialize AES cipher: %w", err)
+	}
+
+	aead, err := cipher.NewGCM(block)
+	if err != nil {
+		return fmt.Errorf("failed to initialize AES-GCM: %w", err)
+	}
+
+	ens.crypt = aead
+	return nil
+}
+
+func (ens *encryptedNonceService) IsStrict() bool    { return true }
+func (ens *encryptedNonceService) IsCrossNode() bool { return false }
+
+func (ens *encryptedNonceService) encryptNonce(counter uint64, expiry time.Time) (token string, err error) {
+	// counter is an 8-byte value and expiry (as a unix timestamp) is
+	// likewise, so we have exactly one block of data.
+	//
+	// We encode in argument order, i.e., (counter, expiry), and in
+	// big endian format.
+
+	// Like Let's Encrypt, we use a 12-byte nonce with the leading four
+	// bytes as zero and the remaining 8 bytes as zeros. This gives us
+	// 2^(8*8)/2 = 2^32 birthday paradox limit to reuse a nonce. However,
+	// note that nonce reuse (in AES-GCM) doesn't leak the key, only the
+	// XOR of the plaintext. Here, as long as they can't forge valid nonces,
+	// we're fine.
+	nonce := make([]byte, 12)
+	for i := 0; i < 4; i++ {
+		nonce[i] = 0
+	}
+	if _, err := io.ReadFull(rand.Reader, nonce[4:]); err != nil {
+		return "", fmt.Errorf("failed to read AEAD nonce: %w", err)
+	}
+
+	plaintext := make([]byte, noncePlaintextLength)
+	binary.BigEndian.PutUint64(plaintext[0:], counter)
+	binary.BigEndian.PutUint64(plaintext[8:], uint64(expiry.Unix()))
+	ciphertext := ens.crypt.Seal(nil, nonce, plaintext, nil)
+
+	// Now, generate the wire format of the nonce. Use a prefix, the nonce,
+	// and then the ciphertext.
+	var wire []byte
```

Comment:
Any reason to throw a version byte/magic on this or anything?

Reply:
We do, `nonceSentinel`, the line below. We verify it on line 170. :-)